x-pack/plugins/watcher/public/sections/watch_edit/components/watch_visualization.tsx

```diff
@@ -0,0 +1,177 @@
+ */
+
+import React, { Fragment, useContext, useEffect, useState } from 'react';
+import '@elastic/charts/dist/style.css';
+import {
+  Axis,
+  CustomSeriesColorsMap,
+  DataSeriesColorsValues,
+  getAxisId,
+  getSpecId,
+  Chart,
+  LIGHT_THEME,
+  LineSeries,
+  Position,
+  ScaleType,
+  Settings,
+} from '@elastic/charts';
+import { debounce } from 'lodash';
+import { TimeBuckets } from 'ui/time_buckets';
+import dateMath from '@elastic/datemath';
+import chrome from 'ui/chrome';
+import moment from 'moment-timezone';
+import { EuiCallOut, EuiSpacer } from '@elastic/eui';
+import { FormattedMessage, injectI18n } from '@kbn/i18n/react';
+import { VisualizeOptions } from 'plugins/watcher/models/visualize_options';
+import { getWatchVisualizationData } from '../../../lib/api';
+import { WatchContext } from './watch_context';
+import { aggTypes } from '../../../models/watch/agg_types';
+
+const getTimezone = () => {
+  const config = chrome.getUiSettingsClient();
+  const DATE_FORMAT_CONFIG_KEY = 'dateFormat:tz';
+  const isCustomTimezone = !config.isDefault(DATE_FORMAT_CONFIG_KEY);
+  if (isCustomTimezone) {
+    return config.get(DATE_FORMAT_CONFIG_KEY);
+  }
+
+  const detectedTimezone = moment.tz.guess();
+  if (detectedTimezone) {
+    return detectedTimezone;
+  }
+
+  const tzOffset = moment().format('Z');
+  return tzOffset;
+};
+
+const getDomain = (watch: any) => {
+  const VISUALIZE_TIME_WINDOW_MULTIPLIER = 5;
+  const fromExpression = `now-${watch.timeWindowSize * VISUALIZE_TIME_WINDOW_MULTIPLIER}${
+    watch.timeWindowUnit
+  }`;
+  const toExpression = 'now';
+  const fromMoment = dateMath.parse(fromExpression);
+  const toMoment = dateMath.parse(toExpression);
+  const visualizeTimeWindowFrom = fromMoment ? fromMoment.valueOf() : 0;
+  const visualizeTimeWindowTo = toMoment ? toMoment.valueOf() : 0;
+  return {
+    min: visualizeTimeWindowFrom,
+    max: visualizeTimeWindowTo,
+  };
+};
+
+const getTimeBuckets = (watch: any) => {
+  const domain = getDomain(watch);
+  const timeBuckets = new TimeBuckets();
+  timeBuckets.setBounds(domain);
+  return timeBuckets;
+};
+
+const loadWatchVisualizationData = debounce(async (watch: any, setWatchVisualizationData: any) => {
+  const domain = getDomain(watch);
+  const timeBuckets = new TimeBuckets();
+  timeBuckets.setBounds(domain);
+  const interval = timeBuckets.getInterval().expression;
+  const visualizeOptions = new VisualizeOptions({
+    rangeFrom: domain.min,
+    rangeTo: domain.max,
+    interval,
+    timezone: getTimezone(),
+  });
+  const { visualizeData } = (await getWatchVisualizationData(watch, visualizeOptions)) as any;
+  const agg = aggTypes[watch.aggType].value;
+  const data = visualizeData[agg];
+  setWatchVisualizationData(data || []);
+}, 500);
+
+const WatchVisualizationUi = () => {
+  const { watch } = useContext(WatchContext);
+  const [watchVisualizationData, setWatchVisualizationData] = useState<number[][]>([]);
```

Comment:
It doesn't have a functional impact, but in case you find it cleaner you could also extract this and the effect on line 95 into a custom hook:

```js
const useWatchVisualizationData = watch => {
  const [watchVisualizationData, setWatchVisualizationData] = useState<number[][]>([]);

  useEffect(
    () => {
      loadWatchVisualizationData(watch, setWatchVisualizationData);
    },
    [watch]
  );

  return watchVisualizationData;
}
```

The consuming it would look like this:

```js
const WatchVisualizationUi = () => {
  const { watch } = useContext(WatchContext);
  const watchVisualizationData = useWatchVisualizationData(watch);
  /* ... */
}
```

Reply:
I think it prefer it in this way personally.